docs/docs/guides/oauth2-grant-type-jwt-bearer.mdx

```diff
@@ -0,0 +1,292 @@
+    "e":"AQAB",
+    "kid":"d8e91f55-67e0-4e56-a066-6a5f0c2efdf7",
+    "n":"nzyis1ZjfNB0bBgKFMSvvkTtwlvBsaJq7S5wA-kzeVOVpVWwkWdVha4s38XM_pa_yr47av7-z3VTmvDRyAHcaT92whREFpLv9cj5lTeJSibyr_Mrm_YtjCZVWgaOYIhwrXwKLqPr_11inWsAkfIytvHWTxZYEcXLgAXFuUuaS3uF9gEiNQwzGTU1v0FqkqTBr4B8nW3HCN47XUu0t8Y0e-lf4s4OxQawWD79J9_5d3Ry0vbV3Am1FtGJiJvOwRsIfVChDpYStTcHTCMqtvWbV6L11BWkpzGXSW4Hv43qa-GSYOD2QU68Mb59oSk2OB-BtOLpJofmbGEGgvmwyCI9Mw"
+  }
+
+  // When this trust relationship expires.
+  "expires_at": "2021-04-23T18:25:43.511Z",
+}
+```
+
+The above example would allow the following JWT Bearer
+
+```
+eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczovL215LWlzc3Vlci5jb20iLCJzdWIiOiJhbGljZUBleGFtcGxlLm9yZyIsImF1ZCI6Imh0dHBzOi8vcHVibGljLmh5ZHJhLmNvbS9vYXV0aDIvdG9rZW4iLCJuYmYiOjEzMDA4MTU3ODAsImV4cCI6MTMwMDgxOTM4MH0.baBiLrVRRU1AKYvAn0X4eIeLvFfpe2wsoD3VTMtaYdbEW2-w-SFeGeEzl5B6sh612bfKkoeihhFVx2md7DP-Rl5asicJzeIhcPETzZbVSPxR1lFdOBwcIPG5N70aSJs2zSn3jnRIhpZf85YZOI8RbQ93Kxla741_4xruHbsNRFqIuWVhxk95BCCnoXzEd8vBTxd_GMn9VijUY_piLPMo-OifRF9pSjYo38aJmRW1tJzeFCMruc9X1W-2c-L_t3rV7zYBH3LlpDZfwyy3T5Pmqf6QKeq1N-MjLnIJcZGT89jqxLmqVFRvAiEyA6iMQXVxmENOnwylGPwuR8DewhWMqg
+```
+
+which has the claims
+
+```json5
+{
+  iss: 'https://my-issuer.com',
+  sub: 'alice@example.org',
+  aud: 'https://public.hydra.com/oauth2/token',
+  nbf: 1300815780,
+  exp: 1300819380
+}
+```
+
+to be exchanged for an OAuth2 Access Token (the `scoe` parameter is optional!)
+
+```
+POST /oauth2/token HTTP/1.1
+Host: public.hydra.com
+Content-Type: application/x-www-form-urlencoded
+
+grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Ajwt-bearer
+&scope=read
+&assertion=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJodHRwczovL215LWlzc3Vlci5jb20iLCJzdWIiOiJhbGljZUBleGFtcGxlLm9yZyIsImF1ZCI6Imh0dHBzOi8vcHVibGljLmh5ZHJhLmNvbS9vYXV0aDIvdG9rZW4iLCJuYmYiOjEzMDA4MTU3ODAsImV4cCI6MTMwMDgxOTM4MH0.baBiLrVRRU1AKYvAn0X4eIeLvFfpe2wsoD3VTMtaYdbEW2-w-SFeGeEzl5B6sh612bfKkoeihhFVx2md7DP-Rl5asicJzeIhcPETzZbVSPxR1lFdOBwcIPG5N70aSJs2zSn3jnRIhpZf85YZOI8RbQ93Kxla741_4xruHbsNRFqIuWVhxk95BCCnoXzEd8vBTxd_GMn9VijUY_piLPMo-OifRF9pSjYo38aJmRW1tJzeFCMruc9X1W-2c-L_t3rV7zYBH3LlpDZfwyy3T5Pmqf6QKeq1N-MjLnIJcZGT89jqxLmqVFRvAiEyA6iMQXVxmENOnwylGPwuR8DewhWMqg
+```
+
+with resulting access token claims:
+
+```
+{
+  "iss": "https://public.hydra.com/",
+  "sub": "alice@example.org",
+  "scp": ["read"],
+  // ...
+}
+```
+
+You can also delete, get, and list trust relationships. Please check the
+[HTTP REST API documentation](../reference/api.mdx) for more details.
+
+### OAuth2 JWT Bearer Grant Type Validation
+
+When performing the `urn:ietf:params:oauth:grant-type:jwt-bearer` Authorization
+Grant, the JWT Bearer in the `assertion` parameter is validated as follows:
+
+1. The JWT MUST contain an `iss` (issuer) claim that contains a unique
+   identifier for the entity that issued the JWT. The value must match the
+   `issuer` value of the trust relationship.
+2. The JWT MUST contain a `sub` (subject) claim identifying the principal that
```

Comment:
Is this a interpretation of the RFC or just how we want to implement this flow in Hydra @aeneasr? Reading [this answer in StackOverflow](https://stackoverflow.com/a/53421198):

> The JWT is issued by a trusted third party (trusted by the authorization server): in this case the subject **could be the client itself, another client or a end user.**

That is what I also understood while reading the RFC as well, see this excerpt:

<img width="569" alt="Captura de pantalla 2021-10-29 a las 19 02 41" src="https://user-images.githubusercontent.com/2232214/139706589-abae21fe-d2eb-4d24-885d-c0e6251758d4.png">

We're very keen on seeing this PR getting merged but if it does while being such restrictive as it's being proposed here, it will limit what kind of use cases could solve.

Keep up the good work!

Reply:
These things can usually be added incrementally. Let's get this merged first and continue from there. If you want to help @Xopek in getting this merged, I am sure he would be happy!

Reply:
I can, but it seems there are others interested in continuing this work too. So, although I would be very happy to get my hands dirty, I prefer to wait until others pronounce!

But in the meantime and in order to have the next steps clear, how would be the best way to proceed @Xopek & @aeneasr?
- Create a new branch on top of your changes and create a new PR?
- Create a new branch and just reference it here so that you can take those changes an apply directly onto @Xopek's branch?
- Should we cleanup the commits before going further? Or we just continue appending commits and wait until we feel that everything is fairly stable?

Thanks in advance!

Reply:
I implemented some end-to-end tests but I'm still not sure how should I proceed to include them in this PR. I started a new branch off of @Xopek's and added my changes: https://github.com/jagobagascon/hydra/commit/b31a67927538e757b4f13aa549d34fc608ecb3a7

There are still some problems: Currently there's no way of requesting a new token without client authentication. Most of the tests are working fine, but I'm still getting the `FOREIGN KEY constraint failed` error when I set `oauth2.grant.jwt.client_auth_optional=true`:

![Screenshot 2021-11-03 at 15 15 26](https://user-images.githubusercontent.com/12712433/140294504-8796dc22-e55c-4c29-91b0-344050a24129.png)

I see several options to fix that:
- Add a default "anonymous" client to the database and use that every time we have an empty client.
- Make the DB column that is failing (`hydra_oauth2_access.client_id`) NULLABLE.
- Or just make client authentication mandatory (this could be a temporary fix) to avoid the problem.

I'm not sure if @Xopek's still available to finish what's pending on this PR, if not I'm open to continue his work.

Reply:
I think best way to proceed is indeed to have separe branch with dedicated commits for e2e tests and etc. So i can cherry pick them on my branch. I am still willing to finish it and will definitely do it. You write so many tests and this is great, i can now focus on other needed changes and try to fix case without client authentication. 

> I implemented some end-to-end tests but I'm still not sure how should I proceed to include them in this PR. I started a new branch off of @Xopek's and added my changes: [jagobagascon@b31a679](https://github.com/jagobagascon/hydra/commit/b31a67927538e757b4f13aa549d34fc608ecb3a7)
> 
> There are still some problems: Currently there's no way of requesting a new token without client authentication. Most of the tests are working fine, but I'm still getting the `FOREIGN KEY constraint failed` error when I set `oauth2.grant.jwt.client_auth_optional=true`:
> 
> ![Screenshot 2021-11-03 at 15 15 26](https://user-images.githubusercontent.com/12712433/140294504-8796dc22-e55c-4c29-91b0-344050a24129.png)
> 
> I see several options to fix that:
> 
>     * Add a default "anonymous" client to the database and use that every time we have an empty client.
> 
>     * Make the DB column that is failing (`hydra_oauth2_access.client_id`) NULLABLE.
> 
>     * Or just make client authentication mandatory (this could be a temporary fix) to avoid the problem.
> 
> 
> I'm not sure if @Xopek's still available to finish what's pending on this PR, if not I'm open to continue his work.

I think best way

Reply:
Great to see things are moving! IMHO the best way to move would be to land this PR by adding as less uncertainty as possible because it already has a lot, even with the e2e. This would also ease the possibility of other collaborating in different aspects of this flow like different `sub` that are different from the clientID or the issue with the client authentication.

So my vote would be for "Or just make client authentication mandatory (**this could be a temporary fix**) to avoid the problem." and address it in a follow-up PR.

Reply:
Sorry, I dropped the ball on this one a bit. I think enforcing client auth right now is sufficient. We can incrementally add improvements to this PR when others want this without client auth!